### PR TITLE
Remove Outbox nonce and checkpointedRoot from dispatch event

### LINKED ISSUE #169: Consider removing Home nonces
These may be a relic of the time when messages needed to be processed in order. It's not clear that the nonces are actually used anywhere...

### PR DESCRIPTION
**Pull request by nambrot**
Fixes #169

A per-destination nonce is unnecessary, and thus this PR removes it. However, some "nonce-like" construct is still necessary to ensure that a message (i.e. the sender/receiver, origin/destination, and the message bytes itself) can have different message hashes over time (since an identical message may be sent multiple times)

I was considering two options:

1) Replace the nonce in the `Message.sol` message with leaf_index.
2) Pass in the leaf_index to the message hash calculation externally.

Pros and Cons:
1) has the nice property that if you just have the message, you can write a method like `message.hash` (in `Message.sol`) that does not need any external information about the message to calculate it's hash. Some folks might assume that the hash of a message should be exclusively over its contents. However, it leads to duplication since the `Dispatch` event (which includes the message), already stores the leaf_index, and thus it would effectively get stored twice in it.

2) makes things slightly more complex as (direct) consumers of the message have to be aware that the message hash of a message needs the leaf_index to be supplied. However, virtually all consumers of abacus messages will gather them from the `Dispatch` event or an abstraction which will have access to the leaf-index. Abstractions can effectively hide this implementation detail so that message classes can actually retain the property of being able to call `message.hash()` See for example `CommittedMessage` (https://github.com/abacus-network/abacus-monorepo/pull/341/files#diff-5b8483436fc79d99cf313336e7c8142a02cf41bab77d42c372d11907cccf6867L77)

Thus I decided for 2)

### CODE REVIEW

File: solidity/core/contracts/Outbox.sol [hunk 1]
```diff
@@ -116,14 +110,10 @@ contract Outbox is IOutbox, Version0, MerkleTreeManager, Common {
         bytes memory _messageBody
     ) external override notFailed returns (uint256) {
         require(_messageBody.length <= MAX_MESSAGE_BODY_BYTES, "msg too long");
-        // get the next nonce for the destination domain, then increment it
-        uint32 _nonce = nonces[_destinationDomain];
-        nonces[_destinationDomain] = _nonce + 1;
         // format the message into packed bytes
         bytes memory _message = Message.formatMessage(
```
Comment (tkporter) [inline solidity/core/contracts/Outbox.sol:114]: Should we be just replacing `_nonce` with `_leafIndex` here? Iirc we still need something nonce-like to ensure there are no message hash conflicts (important because in Inbox.sol we rely on `messages[_messageHash]` for determining if a message has been processed or not), but we just don't need a domain-specific nonce because the leaf index is sufficient to avoid message hash conflicts

File: typescript/hardhat/package.json [hunk 2]
```diff
@@ -16,7 +16,7 @@
     "test": "test"
   },
   "dependencies": {
-    "@abacus-network/core": "^0.0.5",
+    "@abacus-network/core": "^0.0.6",
```
Comment (tkporter) [inline typescript/hardhat/package.json:19]: was this needed?

Comment (nambrot) [reply typescript/hardhat/package.json:19]: The API changes for [TestAbacusDeploy.ts](https://github.com/abacus-network/abacus-monorepo/pull/341/files#diff-ca6b384d7e6dd7ed8ff62341a817062118ec47b524806459f49d3b6c5bc02ff6). Packages in the monorepo should depend on the latest package version of other dependencies in the repo, and someone missed it here when bumping core before

Comment (asaj) [reply typescript/hardhat/package.json:19]: I wonder who that "someone" was 🙃

File: solidity/core/contracts/Outbox.sol (hunk 1 above)
Comment (nambrot) [reply solidity/core/contracts/Outbox.sol:116]: Like discussed on Discord, great point. I ultimately decided to go with explicitly passing the leaf_index to the message hash calculation instead of embedding the leaf_index inside the message, because the duplication felt very wrong, and most places that deal with messages retrieve them from the dispatch even where they will already have access to the leafindex

File: rust/abacus-core/src/traits/message.rs [hunk 3]
```diff
@@ -21,7 +20,15 @@ impl RawCommittedMessage {
     /// The leaf is the keccak256 digest of the message, which is committed
     /// in the message tree
     pub fn leaf(&self) -> H256 {
-        keccak256(&self.message).into()
+        let buffer = [0u8; 28];
+        H256::from_slice(
+            Keccak256::new()
```
Comment (tkporter) [inline rust/abacus-core/src/traits/message.rs:25]: I'm a little wary of doing any bespoke formatting of the contents here. Does the chaining here for sure always do the exact same thing as `abi.encodePacked`? The documentation on `abi.encodePacked` [here](https://docs.soliditylang.org/en/v0.8.13/abi-spec.html#non-standard-packed-mode) has me feeling like this is doing something different? If there's some documentation that explains the chaining here is correct it'd probably be good to leave it in a comment

To be fair I can't find anything in ethers-rs that does `abi.encodePacked`, just `abi.encode` from what I can see...

File: rust/abacus-core/src/types/messages.rs [hunk 4]
```diff
@@ -75,30 +70,28 @@ impl Decode for AbacusMessage {
...
 impl AbacusMessage {
     /// Convert the message to a leaf
-    pub fn to_leaf(&self) -> H256 {
-        keccak256(self.to_vec()).into()
-    }
-
-    /// Get the encoded destination + nonce
-    pub fn destination_and_nonce(&self) -> u64 {
-        utils::destination_and_nonce(self.destination, self.nonce)
+    pub fn to_leaf(&self, leaf_index: u32) -> H256 {
```
Comment (tkporter) [inline rust/abacus-core/src/types/messages.rs:80]: ditto here on ensuring we are replicating `abi.encodePacked` correctly- also, if we do stay with this solution, any way we can pull the contents into a function that this `to_leaf` and `RawCommittedMessage`'s `to_leaf` can use?

File: rust/config/test/alfajores_config.json [hunk 5]
```diff
@@ -49,14 +37,26 @@
...
+      "url": ""
+    }
+  },
   "tracing": {
     "level": "debug",
     "fmt": "json"
   },
   "db": "db_path"
-}
+}
```
Comment (tkporter) [inline rust/config/test/alfajores_config.json:62]: nit: I guess keep trailing newline?

File: rust/tools/abacus-cli/src/subcommands/db_state.rs [hunk 6]
```diff
@@ -44,41 +44,23 @@ impl DbStateCommand {
         Ok(())
     }
 
-    fn create_comitted_root_to_message_map(
-        &self,
-        db: &AbacusDB,
-    ) -> Result<HashMap<H256, Vec<CommittedMessage>>> {
-        let mut messages_by_committed_roots: HashMap<H256, Vec<CommittedMessage>> = HashMap::new();
+    fn create_comitted_root_to_message_map(&self, db: &AbacusDB) -> Result<Vec<CommittedMessage>> {
```
Comment (tkporter) [inline rust/tools/abacus-cli/src/subcommands/db_state.rs:47]: is this name of the fn accurate? also curious why the change here

File: rust/tools/abacus-cli/src/subcommands/db_state.rs [hunk 7]
```diff
@@ -89,7 +71,7 @@ impl DbStateCommand {
     fn create_output_vec(
         &self,
         _db: &AbacusDB,
-        _messages_by_committed_roots: HashMap<H256, Vec<CommittedMessage>>,
+        _messages_by_committed_roots: Vec<CommittedMessage>,
```
Comment (tkporter) [inline rust/tools/abacus-cli/src/subcommands/db_state.rs:74]: ditto is this name accurate

**Review by tkporter — COMMENTED** [review]
Looking for some confirmation that the rust `abi.encodePacked` is accurate

File: rust/abacus-core/src/traits/message.rs (hunk 3 above)
Comment (nambrot) [reply rust/abacus-core/src/traits/message.rs:25]: What exactly do you mean by bespoke here? I just copied this from how we hash the digest for the checkpoints as well https://github.com/abacus-network/abacus-monorepo/blob/nambrot/multiple-inboxes-per-domain/rust/abacus-core/src/types/checkpoint.rs#L72:L72

While I don't think I have hard evidence, certainly the consistent use of this function makes me pretty confident that it works as intended. I think what might be confusing is the "buffer" of 28 0s. The buffer is not for ABI encoding, but the fact that on-chain, the leaf_index is a uint256, while in rust land it is a u32. Therefore we need to pad the leaf_index so that the pre-image is the same. Otherwise at least it seems to me that both `Keccak256::new.chain` and `abi.encodePacked`, basically just pack the bytes together

File: rust/abacus-core/src/types/messages.rs (hunk 4 above)
Comment (nambrot) [reply rust/abacus-core/src/types/messages.rs:80]: How strongly do you feel about this? IMO, this level of duplication isn't worth the redirection of extraction

File: rust/tools/abacus-cli/src/subcommands/db_state.rs (hunk 6 above)
Comment (nambrot) [reply rust/tools/abacus-cli/src/subcommands/db_state.rs]: Oh I just made the change to make the compiler happy but basically this command no longer makes any sense for us, so I just removed this for now, and we can re-add once we find it is useful for us specifically

File: rust/abacus-core/src/traits/message.rs (hunk 3 above)
Comment (tkporter) [reply rust/abacus-core/src/traits/message.rs:25]: ahhhh my bad, I didn't realize this was already present somewhere. By bespoke I meant circumventing any function that may exist already that matches abi.encodePacked behavior (like in ethers-rs or something). I'm ok with this implementation, thanks for the info

File: rust/abacus-core/src/types/messages.rs (hunk 4 above)
Comment (tkporter) [reply rust/abacus-core/src/types/messages.rs:80]: not too strongly so happy to keep as is

**Review by tkporter — APPROVED** [review]
